src/encoding/proto.rs

```diff
@@ -0,0 +1,839 @@
+//! # let mut registry = Registry::default();
+//! # let counter: Counter = Counter::default();
+//! # registry.register(
+//! #   "my_counter",
+//! #   "This is my counter",
+//! #   counter.clone(),
+//! # );
+//! # counter.inc();
+//! // Returns `MetricSet`, the top-level container type. Please refer to [openmetrics_data_model.proto](https://github.com/OpenObservability/OpenMetrics/blob/main/proto/openmetrics_data_model.proto) for details.
+//! let metric_set = encode(&registry);
+//!
+//! let family = metric_set.metric_families.first().unwrap();
+//! assert_eq!("my_counter", family.name);
+//! assert_eq!("My counter.", family.help);
+//! ```
+
+// Include the `openmetrics_data_model` module, which is generated from `proto/openmetrics_data_model.proto`.
+pub mod openmetrics_data_model {
+    include!(concat!(env!("OUT_DIR"), "/openmetrics.rs"));
+}
+
+use crate::metrics::counter::Counter;
+use crate::metrics::exemplar::{CounterWithExemplar, Exemplar, HistogramWithExemplars};
+use crate::metrics::family::{Family, MetricConstructor};
+use crate::metrics::gauge::Gauge;
+use crate::metrics::histogram::Histogram;
+use crate::metrics::info::Info;
+use crate::metrics::{counter, gauge, MetricType, TypedMetric};
+use crate::registry::Registry;
+use std::collections::HashMap;
+use std::ops::Deref;
+
+pub fn encode<M>(registry: &Registry<M>) -> openmetrics_data_model::MetricSet
+where
+    M: EncodeMetric,
+{
+    let mut metric_set = openmetrics_data_model::MetricSet::default();
+
+    for (desc, metric) in registry.iter() {
+        let mut family = openmetrics_data_model::MetricFamily::default();
+        family.name = desc.name().to_string();
+        family.r#type = {
+            let metric_type: openmetrics_data_model::MetricType = metric.metric_type().into();
+            metric_type as i32
+        };
+        if let Some(unit) = desc.unit() {
+            family.unit = unit.as_str().to_string();
+        }
+        family.help = desc.help().to_string();
+        family.metrics = metric.encode(desc.labels().encode());
+
+        metric_set.metric_families.push(family);
+    }
+
+    metric_set
+}
+
+impl From<MetricType> for openmetrics_data_model::MetricType {
+    fn from(m: MetricType) -> Self {
+        match m {
+            MetricType::Counter => openmetrics_data_model::MetricType::Counter,
+            MetricType::Gauge => openmetrics_data_model::MetricType::Gauge,
+            MetricType::Histogram => openmetrics_data_model::MetricType::Histogram,
+            MetricType::Info => openmetrics_data_model::MetricType::Info,
+            MetricType::Unknown => openmetrics_data_model::MetricType::Unknown,
+        }
+    }
+}
+
+/// Trait implemented by each metric type, e.g. [`Counter`], to implement its encoding.
+pub trait EncodeMetric {
+    fn encode(
+        &self,
+        labels: Vec<openmetrics_data_model::Label>,
+    ) -> Vec<openmetrics_data_model::Metric>;
```

Comment:
Disclaimer: I haven't fully thought this through.

Would it be an option to return an iterator (e.g. `Self::Iterator`) here? That would eliminate the many `vec!` allocations and would only require a single final `collect`.

Reply:
Note to self: [Returning Rust Iterators | Depth-First](https://depth-first.com/articles/2020/06/22/returning-rust-iterators/)

Reply:
Note to self: 

- I tried `Associated Type` which is described [here](https://depth-first.com/articles/2020/06/22/returning-rust-iterators/#associatedtype).
- Commit log: https://github.com/ackintosh/client_rust/commit/e4c46dfcd25eba32801229fb3d43e7fb9ca08338 (on another branch `protobuf-support-associated-type`)
- Build error below occurs:

```text
error[E0207]: the lifetime parameter `'a` is not constrained by the impl trait, self type, or predicates
   --> src/encoding/proto.rs:199:6
    |
199 | impl<'a, N, A> EncodeMetric for Counter<N, A>
    |      ^^ unconstrained lifetime parameter
```

Reply:
Next, I will try [`Simple Iterator Delegation`](https://depth-first.com/articles/2020/06/22/returning-rust-iterators/#simpleiteratordelegation).

Reply:
@ackintosh thanks for investigating. Let me know in case you want me to take a look as well.

Reply:
I have fixed the [build error](https://github.com/prometheus/client_rust/pull/47#discussion_r933917283) and now it works well (tests are green).

Here is a PR: https://github.com/ackintosh/client_rust/pull/2

However, the allocation count seems not reduced as far as I checked. 🤔 [Here](https://github.com/ackintosh/client_rust/pull/2#issuecomment-1203634863) is a detail of how I checked the allocation count.

@mxinden Please take a look at https://github.com/ackintosh/client_rust/pull/2 and let me know if you have any concerns. 🙏

Reply:
@ackintosh is https://github.com/ackintosh/client_rust/pull/5 of any help?

Reply:
Thank you for the patch @mxinden ! I have learned a lot. ✨ Will merge later.

Reply:
Wonderful. Ping me in case I should do another review or in case you need any help.